app/controllers/api/custom_buttons_controller.rb

```diff
@@ -2,32 +2,30 @@ module Api
   class CustomButtonsController < BaseController
     def create_resource(_type, _id, data)
       CustomButton.transaction do
-        custom_button = CustomButton.new(data.except("resource_action", "options"))
-        custom_button.userid = User.current_user.userid
-        custom_button.options = data["options"].deep_symbolize_keys if data["options"]
-        custom_button.resource_action.update!(data["resource_action"].deep_symbolize_keys) if data.key?("resource_action")
-        custom_button.save!
-        custom_button
+        CustomButton.new(data.except("resource_action", "options")).tap do |cb|
+          cb.userid = User.current_user.userid
+          cb.options = data["options"].deep_symbolize_keys if data["options"]
+          cb.create_resource_action!(data["resource_action"].deep_symbolize_keys) if data.key?("resource_action")
+          cb.save!
+        end
       end
-    rescue
-      raise BadRequestError, "Failed to create new custom button - #{custom_button.errors.full_messages.join(", ")}"
+    rescue => err
+      raise BadRequestError, "Failed to create new custom button - #{err.message}"
     end
 
     def edit_resource(type, id, data)
       return if data.empty?
 
       CustomButton.transaction do
-        custom_button = fetch_custom_button(type, id)
-        updated_data = data['resource'] || data
-        if updated_data['resource_action'].present?
-          custom_button.resource_action = create_or_update_resource_action(updated_data['resource_action'])
+        fetch_custom_button(type, id).tap do |cb|
+          updated_data = data['resource'] || data
+          cb.resource_action = create_or_update_resource_action(updated_data['resource_action']) if updated_data['resource_action'].present?
+          updated_data.except!('resource_action')
+          cb.update!(updated_data.deep_symbolize_keys) if updated_data.present?
```

Comment:
```suggestion
          cb.update!(updated_data.deep_symbolize_keys)
```

I think we always want to call update
And `if !updated_data.present?` then updated_data will be empty and the attributes will be a no-op but we want to save for the create_or_update action.

Are we sure that a save to this button also saves resource_action? (I always get confused between `has_one` and `belongs_to`. One of them saves and the other does not.

Reply:
Yeah, I agree with this suggestion, and yes, we're pretty definite this does also save the resource action.